pkg/gridtypes/zos/ipv4.go

```diff
@@ -1,36 +1,81 @@
 package zos
 
 import (
+	"fmt"
 	"io"
 	"net"
 
 	"github.com/threefoldtech/zos/pkg/gridtypes"
 )
 
-// PublicIP structure
-type PublicIP struct{}
+// PublicIP4 structure
+// this is a deprecated type and only kept here for backward compatibility with
+// older deployments.
+// Please use PublicIP instead
+type PublicIP4 struct{}
+
+// Valid validate public ip input
+func (p PublicIP4) Valid(getter gridtypes.WorkloadGetter) error {
+	return nil
+}
+
+// Challenge implementation
+func (p PublicIP4) Challenge(b io.Writer) error {
+	return nil
+}
+
+// Capacity implementation
+func (p PublicIP4) Capacity() (gridtypes.Capacity, error) {
+	return gridtypes.Capacity{IPV4U: 1}, nil
+}
+
+type PublicIP struct {
+	// V4 use one of the reserved Ipv4 from your contract. The Ipv4
+	// itself costs money + the network traffic
+	V4 bool `json:"v4"`
+	// V6 get an ipv6 for the VM. this is for free
+	// but the consumed capacity (network traffic) is not
+	V6 bool `json:"v6"`
+}
 
 // Valid validate public ip input
 func (p PublicIP) Valid(getter gridtypes.WorkloadGetter) error {
+	if !p.V4 && !p.V6 {
+		return fmt.Errorf("public ip workload with no selections")
+	}
+
 	return nil
 }
 
 // Challenge implementation
-func (p PublicIP) Challenge(b io.Writer) error {
+func (p PublicIP) Challenge(w io.Writer) error {
+	if _, err := fmt.Fprintf(w, "%t", p.V4); err != nil {
+		return err
+	}
+
+	if _, err := fmt.Fprintf(w, "%t", p.V6); err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // Capacity implementation
 func (p PublicIP) Capacity() (gridtypes.Capacity, error) {
-	return gridtypes.Capacity{IPV4U: 1}, nil
+	var c uint64
+	if p.V4 {
+		c = 1
+	}
+	return gridtypes.Capacity{IPV4U: c}, nil
```

Comment:
The capacity calculation here indicates that the user is billed only on active PublicIP workloads. So if a user made a contract with an ip and didn't use it in any deployment, would he be billed?

Reply:
Yes, he will. the billing of the reserved IPv4 is handled by the contract itself. I think we have this here just for the completeness